docs/syntax.md

```diff
@@ -0,0 +1,512 @@
+- A variable.
+- A constructor.
+- A tuple.
+- A superposition.
+
+The rule body is a `term`, there is no statements in the language.
+
+### Datatype
+
+Defines an Algebraic Data Type, it should have at least one constructor.
+
+```rust
+data Tree
+  = (Leaf value)
+  | (Node left right)
+  | Nil
+```
+
+`Tree` is the ADT name and it should be unique, except that it can be used once by a constructor name.
+
+Each constructor is defined by a name followed by its fields.
+
+## Terms
+
+### Lambda
+
+```rust
+@x x
+
+λx x
+
+λ(fst, snd) snd
+
+λ{x y} x
+```
+
+Lambdas represents functions, it can be written with `λ` or `@` followed by a pattern and a term.
+
+A pattern is equivalent to a let:
+
+```rust
+λa let (fst, snd) = a; snd
+
+λa let {x y} = a; x
+```
+
+### Unscoped Lambdas and Variables
+
+```
+λ$x $x
+```
+
+Same as lambdas, with the exception that the variable starts with a `$` sign.
+
+Read [using scopeless lambdas](/docs/using-scopeless-lambdas.md) to know more about.
+
+### Application
+
+```rust
+(fun argn)
+```
+
+An application is surrounded by `(` `)`, written in lisp style.
+
+> Lambdas have a higher precedence, so `(@x x 1)` and `((@x x) 1)` means the same thing.
+
+### Tuples
+
+```rust
+(1, 2, 3)
+```
+
+A tuple is surrounded by `(` `)`, with the difference that it's elements are separated by `,`.
+
+### Superposition
+
+```rust
+{1 2 3}
+```
+
+A superposition of values is defined using `{` `}` with at least 2 terms inside.
+
+### Let-bindings
+
+```rust
+let x = (+ 1 2)
+x
+
+let (fst, snd, era) = (1, 2, *);
+(+ fst snd)
+
+let {f1 f2} = λx x;
+(f1 f2)
+
+let $x = (some_fn $x);
+*
+```
+
+> `*` is an eraser term.
+
+A let term uses a pattern, it can be:
+- A variable / unscoped variable.
+- A tuple.
+- A superposition.
+
+The let term will expects a binding value followed by a `next` term.
+
+Using `;` is optional.
+
+### Switch
+
+```rust
+switch n {
+  0: "zero"
+  1: "one"
+  _: "greater than 1"
+}
+
+switch x = (+ 1 1) {
+  0: 42;
+  _: x-1;
+}
+```
+
+A switch for native numbers, it can hold a name binding if the matching term is not a variable.
+
+The cases need to be typed from `0` to a wildcard `_` in sequence.
+
+Using `;` is optional.
+
+### Match
+
+```rust
+match opt = (Some "Bend") {
+  Some: opt.value;
+  None: "No name";
+}
+```
+
+A pattern match expression, it can hold a name binding if the matching term is not a variable.
+
+It is possibile to use a _wildcard_ variable or `*` as an exhaustive default case.
+
+Using `;` is optional.
+
+### Monadic bind
+
+```rust
+Result.bind (Result.ok val) f = (f val)
+Result.bind err _ = err
+
+div a b = switch b {
+  0: (Result.err "Div by 0")
+  _: (Result.ok (/ a b))
+}
+
+rem a b = switch b {
+  0: (Result.err "Mod by 0")
+  _: (Result.ok (% a b))
+}
+
+Main = do Result.bind {
+  ask y = (div 3 2);
+  ask x = (rem y 0);
+  x
+}
+```
+
+Receives a monadic bind function and then expects a block.
+
+The block is followed by `ask` binds and ends with a return term.
```

Comment:
Add that the bind function should be of type `(Monad a) -> (a -> (Monad b)) -> (Monad b)`.

Add that it forwards any used variables and point to the doc (which we still have to write)